File: Codeforces/gym/2020uncpc/a.py
from math import *
from fractions import *

def sqr(x):
  return x * x

class Point:
  def __init__(self, x = Fraction(0), y = Fraction(0)):
    self.x = x
    self.y = y

  def read(self):
    a, b = map(int, input().split(' '))
    self.x = Fraction(a)
    self.y = Fraction(b)

  def abs(self):
    return sqrt(self.x * self.x + self.y * self.y)

  def distance(self, b):
    x = self.x - b.x
    y = self.y - b.y
    return sqrt(x * x + y * y)

  def unit(self):
    l = self.abs()
    return Point(self.x / l, self.y / l)

def cal(a, b, c, L, R):
  eval = lambda x : a * x * x + b * x + c

  ans = min(eval(L), eval(R))

  if a != 0:
    mid = -b / (a + a)
    if L <= mid and mid <= R:
      ans = min(ans, eval(mid))

  if ans <= 0:
    ans = 0
  
  return ans

x1, x2, x3, x4 = map(int, input().split(' '))
y1, y2, y3, y4 = map(int, input().split(' '))

a = Point(Fraction(x1), Fraction(x2))
b = Point(Fraction(x3), Fraction(x4))
c = Point(Fraction(y1), Fraction(y2))
d = Point(Fraction(y3), Fraction(y4))

ab = a.distance(b)
cd = b.distance(d)

if Point(b.x - a.x, b.y - a.y).abs() == 0 and Point(d.x - c.x, d.y - c.y).abs() == 0:
  print(a.distance(c))
if Point(b.x - a.x, b.y - a.y).abs() == 0:
  k2 = Point(d.x - c.x, d.y - c.y).unit()
  A = sqr(k2.x) + sqr(k2.y)
  B = 2.0 * (c.x - b.x) * k2.x + 2.0 * (c.y - b.y) * k2.y
  C = sqr(c.x - b.x) + sqr(c.y - b.y)
  print(sqrt(cal(A, B, C, ab, cd)))
elif Point(d.x - c.x, d.y - c.y).abs() == 0:
  k1 = Point(b.x - a.x, b.y - a.y).unit()
  A = sqr(k1.x) + sqr(k1.y)
  B = 2.0 * (a.x - d.x) * k1.x + 2.0 * (a.y - d.y) * k1.y
  C = sqr(a.x - d.x) + sqr(a.y - d.y)
  print(sqrt(cal(A, B, C, cd, ab)))
else:
  k1 = Point(b.x - a.x, b.y - a.y).unit()
  k2 = Point(d.x - c.x, d.y - c.y).unit()

  A = sqr(k1.x - k2.x) + sqr(k1.y - k2.y)
  B = 2.0 * (k1.x - k2.x) * (a.x - c.x) + 2.0 * (k1.y - k2.y) * (a.y - c.y)
  C = sqr(a.x - c.x) + sqr(a.y - c.y)
  ans = cal(A, B, C, 0, min(ab, cd))

  if ab < cd:
    A = sqr(k2.x) + sqr(k2.y)
    B = 2.0 * (c.x - b.x) * k2.x + 2.0 * (c.y - b.y) * k2.y
    C = sqr(c.x - b.x) + sqr(c.y - b.y)
    ans = min(ans, cal(A, B, C, ab, cd))
  elif ab > cd:
    A = sqr(k1.x) + sqr(k1.y)
    B = 2.0 * (a.x - d.x) * k1.x + 2.0 * (a.y - d.y) * k1.y
    C = sqr(a.x - d.x) + sqr(a.y - d.y)
    ans = min(ans, cal(A, B, C, cd, ab))

  print(sqrt(ans))

File: Codeforces/gym/2020uncpc/a.cpp
#include <iostream>
#include <cstdio>
#include <cassert>
#include <cstring>
#include <cmath>
#include <functional>
#include <algorithm>
#include <utility>
#include <vector>
#include <string>
#include <map>
#include <set>
#ifdef XLor
  #define dbg(args...) cout << "\033[32;1m" << #args << " -> ", err(args)
  void err() { std::cout << "\033[39;0m" << std::endl; }
  template<typename T, typename...Args>
  void err(T a, Args...args) { std::cout << a << ' '; err(args...); }
#else
  #define dbg(...)
#endif
#define ms(a,b) memset(a,b,sizeof(a))
using namespace std;
using ll = long long;
using PII = pair<int,int>;
const int mod = 998244353;
const int inf = 1 << 30;
const int maxn = 200000 + 5;

namespace Geometry {
  const long double eps = 1e-14;
  const long double pi = acos(-1.0);

  inline int sign(long double x) {
    if (fabs(x) < eps) return 0;
    else return x < 0 ? -1 : 1;
  }
  inline int dcmp(long double x, long double y) {
    return sign(x - y);
  }
  inline int zero(long double x) {
    return fabs(x) < eps;
  }

  struct Point {
    long double x, y;
    Point(long double x = 0, long double y = 0): x(x), y(y) {}

    void read() {
      int a, b;
      scanf("%d%d", &a, &b);
      x = a;
      y = b;
    }

    Point operator + (const Point& b) const { return Point(x + b.x, y + b.y); }
    Point operator - (const Point& b) const { return Point(x - b.x, y - b.y); }
    Point operator * (long double p) { return Point(x * p, y * p); }
    Point operator / (long double p) { return Point(x / p, y / p); }

    bool operator == (const Point& b) const { return dcmp(x, b.x) == 0 && dcmp(y, b.y) == 0; }
    bool operator < (const Point& b) const {
      int f = dcmp(x, b.x);
      if (f == -1) return 1;
      else if (f == 1) return 0;
      else return dcmp(y, b.y) == -1;
    }

    long double abs() { return sqrt(x * x + y * y); }
    long double abs2() { return x * x + y * y; }
    long double distance(const Point& b) { return (*this - b).abs(); }

    Point unit() { return *this / abs(); }
  };

  ostream& operator<<(ostream& out, const Point& u) {
    return out << "(" << u.x << ", " << u.y << ")";
  }
  istream& operator>>(istream& in, Point& u) {
    return in >> u.x >> u.y;
  }
}
using namespace Geometry;

Point a, b, c, d;

inline long double sqr(long double x) {
  return x * x;
}

long double cal(long double L, long double R, function<long double(long double)> eval) {
  long double ans = min(eval(L), eval(R));
  long double l = L, r = R;
  for (int i = 0; i < 20000; i++) {
    long double ml = l + (r - l) / 3.0;
    long double mr = r - (r - l) / 3.0;
    ans = min(ans, eval(ml));
    ans = min(ans, eval(mr));
    if (eval(ml) > eval(mr)) {
      l = ml;
    } else {
      r = mr;
    }
  }
  if (sign(ans) <= 0) {
    return 0.0;
  }
  return ans;
}

int main() {
  a.read();
  b.read();
  c.read();
  d.read();
  
  Point k1, k2;
  if (a == b) k1.x = k1.y = 0;
  else k1 = (b - a).unit();
  if (c == d) k2.x = k2.y = 0;
  else k2 = (d - c).unit();

  long double ab = a.distance(b), cd = c.distance(d);

  long double ans = cal(0, min(ab, cd), [&](long double t) {
    long double dx = (a.x + k1.x * t) - (c.x + k2.x * t);
    long double dy = (a.y + k1.y * t) - (c.y + k2.y * t);
    return dx * dx + dy * dy;
  });

  if (dcmp(ab, cd) < 0) {
    ans = min(ans, cal(ab, cd, [&](long double t) {
      long double dx = b.x - (c.x + k2.x * t);
      long double dy = b.y - (c.y + k2.y * t);
      return dx * dx + dy * dy;
    }));
  } else if (dcmp(ab, cd) > 0) {
    ans = min(ans, cal(cd, ab, [&](long double t) {
      long double dx = (a.x + k1.x * t) - d.x;
      long double dy = (a.y + k1.y * t) - d.y;
      return dx * dx + dy * dy;
    }));
  }

  printf("%.10Lf\n", sqrt(ans));

  return 0;
}